Add Ethernet support with W5500 integration

tototo31 opened the pull request:
Updated sdkconfig to enable Ethernet support and configured W5500 pins. Introduced ethernet_manager for initialization, connection status checks, and IP information retrieval. Updated CMakeLists and Kconfig for Ethernet integration. Enhanced command handling for bringing Ethernet up and down.

### **be sure to also enable the ethernet component in component config -> ethernet  -> support SPI to ethernet module AND in Ghost esp options**

### **also need to enable component config -> LWIP -> Enable LWIP statistics if we want stats to work**


# What's new (Author - fill this out)
-  support for w5500 ethernet
- added: `ethup` and `ethdown` commands
- Added `etharp` for arp scans
- added `ethports` for tcp port scans (`etharp <ip> <port>`) (also supports "`local`" as an ip to scan the gateway and "`all`" as a port argument to scan all ports) 
- Added `ethping` to perform a ping scan
- Added `ethdns` for dns resolution and lookup
- added `ethtrace` for traceroute
- added `ethstats` for statisttics
- `ethconfig` `dhcp` or `static <ip> <netmask> <gateway>` or `ethconfig show `
- `ethmac ` to manage mac address
- `ethserv <ip>` for service discovery/banner grabbing
- `ethntp <time_server>` for ntp time syncing the system clock. pool.ntp.org is the default if no server is provided
- `ethhttp https://<url> <lines to show|all>` to perform get requests against a domain
- `time` and `settime <unix_timestamp>` commands for getting and setting system time

# Changed
- added pindefs in kconfig for w5500 spi pins

# Removed
-

# Verification (Author - fill this out)
- I have tested Primary device(s) (list exact HW + config): something 2
- I have tested potentially affected devices (and/or list potential affected devices not available to you):  
- I have wrapped device specific code with `#ifdef CONFIG_...` or similar: [x] Yes / [ ] N/A
- I have updated Hugo Docs with any new or changed info for end users: [x] Yes / [ ] N/A

## Linked Issues
- Closes #

## Checklist (Reviewer - don't fill this out)
- [ ] Code builds, flashes, and feature verified with no functional issues on listed device(s)
- [ ] Changes reviewed for unintended impact on other devices/targets


<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Introduce Ethernet (W5500) support with a new manager, extensive CLI/network tools, and an option to restrict WebUI to AP clients.
> 
> - **Networking/Ethernet**:
>   - Add `ethernet_manager` (W5500 over SPI) with link/IP/DHCP handling and netif glue.
>   - New helpers: `eth_utils`, `eth_http`, and `eth_fingerprint` (mDNS/NetBIOS/SSDP discovery).
>   - Initialize Ethernet in `main.c` (conditional) and include `esp_eth`.
> - **CLI/Features**:
>   - New commands: `ethup`, `ethdown`, `ethinfo`, `etharp`, `ethports`, `ethping`, `ethdns`, `ethtrace`, `ethstats`, `ethconfig`, `ethmac`, `ethserv`, `ethntp`, `ethhttp`.
>   - Add `time` and `settime` commands; help updated; dual-comm command routing supports Ethernet ops.
> - **Settings/Web**:
>   - Add setting `webui_restrict_to_ap` (NVS persisted) with getters/setters and `webuiap` command.
>   - Gate all WebUI/API handlers to AP subnet when enabled; add Ethernet help/quick actions in site UI; improve terminal inputs.
> - **UI (Options/DualComm)**:
>   - Add Ethernet menu (info, scans, ports, DNS, trace, HTTP, NTP, stats, config) and WebUI AP-only toggle.
> - **Build/Config**:
>   - CMake: add `esp_eth` component and conditional `ethernet_manager.c`.
>   - Kconfig: `WITH_ETHERNET` and W5500 pin configs (MOSI/MISO/SCK/CS/INT).
> - **Misc**:
>   - Minor NFC/desfire comment tweak; stop routines cover Ethernet scans; docs/help pages updated.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit ddd7d72ac1b9bc2cd182b6b32cf71b4419dd6a60. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

Commit:
Add Ethernet support with W5500 integration

Updated sdkconfig to enable Ethernet support and configured W5500 pins. Introduced ethernet_manager for initialization, connection status checks, and IP information retrieval. Updated CMakeLists and Kconfig for Ethernet integration. Enhanced command handling for bringing Ethernet up and down.

Commit:
Update CMakeLists.txt to include esp_eth in required components for Ethernet support for now

Commit:
Add Ethernet Manager enhancements for DNS and DHCP information retrieval

- Introduced functions to retrieve DNS server information and DHCP server IP address.
- Updated command handling to display Ethernet status, IP information, DNS servers, and DHCP server details.
- Enhanced logging for better visibility of network status and configuration.
- Added necessary includes and improved error handling in Ethernet manager implementation.

Commit:
Add ARP scanning command for Ethernet support

- Implemented `handle_eth_arp_cmd` to perform ARP scans on the connected Ethernet network.
- Added error handling for Ethernet connection status, IP information retrieval, and LWIP netif access.
- Registered the new command in the command handling system for user access.
- Enhanced logging to provide detailed output during the ARP scanning process.

Commit:
Add Ethernet port scanning command

- Implemented `handle_eth_ports_cmd` to scan TCP ports on a specified IP address.
- Added argument parsing for local network scanning and custom port ranges.
- Enhanced logging to provide progress updates and results of the port scan.
- Registered the new command in the command handling system for user access.

Commit:
Add Ethernet ping scanning command

- Implemented `handle_eth_ping_cmd` to perform ICMP ping scans on the connected Ethernet network.
- Added a local definition for the ICMP packet structure and checksum calculation.
- Enhanced logging to provide progress updates and results of the ping scan.
- Registered the new command in the command handling system for user access.

Commit:
Enhance logging for network scan commands

- Updated `handle_eth_arp_cmd`, `handle_eth_ports_cmd`, and `handle_eth_ping_cmd` to provide detailed summaries of scan results.
- Added information on scanned hosts, success rates, and timeout settings for better user feedback.
- Improved formatting of log messages for clarity and consistency across different scan types.

Commit:
Add Ethernet command enhancements for DNS, traceroute, and statistics

- Implemented `handle_eth_dns_cmd` for DNS resolution and reverse lookup, with detailed usage instructions and error handling.
- Added `handle_eth_trace_cmd` for performing traceroute operations, including hostname resolution and ICMP echo requests.
- Introduced `handle_eth_stats_cmd` to display Ethernet statistics, including link status, IP configuration, and packet statistics.
- Registered new commands in the command handling system for user access.
- Enhanced logging for improved user feedback and clarity in command outputs.

Commit:
Update command-line reference to include Ethernet commands and enhance help documentation

- Added new Ethernet commands for connection management, network scanning, network tools, configuration, and statistics.
- Updated the help command to include Ethernet as a category for better user guidance.
- Enhanced documentation clarity and structure for improved usability.

Commit:
Enhance command-line help for Ethernet functionality

- Added detailed help documentation for new Ethernet commands, including connection management, network scanning, and configuration.
- Updated the command-line help system to include Ethernet as a category, improving user guidance and accessibility.
- Enhanced overall clarity and structure of help messages for better usability.

tototo31 committed:
Merge branch 'Development-deki' into ethernet-support

Commit:
Add Ethernet menu options and functionality

- Introduced a new Ethernet menu in the options screen, allowing users to access Ethernet-related commands.
- Added options for Ethernet management, including "Ethernet Up", "Ethernet Down", "Ethernet Info", and various network scanning tools.
- Implemented callback functions for DNS lookup and traceroute operations, enhancing network diagnostics capabilities.
- Updated the options menu to handle the new Ethernet state and commands, improving user navigation and functionality.

Commit:
Add Ethernet command options and update .gitignore

- Added new Ethernet command options in the help documentation and user interface, including commands for managing Ethernet connections and performing network scans.
- Updated the .gitignore file to include the 'venv' directory, ensuring it is excluded from version control.

Commit:
Update ghost_site.html GZ data size and content

- Increased the size of the compressed HTML data for ghost_site.html from 20701 to 21887 bytes.
- Updated the corresponding byte array to reflect the new content, ensuring the latest version of the HTML is included in the project.

tototo31 commented:
added help wanted tag for the cmake bs and testing

jaylikesbunda commented on configs/sdkconfig.somethingsomething2:
```diff
@@ -1933,16 +1939,15 @@ CONFIG_LOG_VERSION=1
-CONFIG_LOG_DEFAULT_LEVEL_WARN=y
-# CONFIG_LOG_DEFAULT_LEVEL_INFO is not set
+# CONFIG_LOG_DEFAULT_LEVEL_WARN is not set
+CONFIG_LOG_DEFAULT_LEVEL_INFO=y
```
just noting we don't wanna change default log levels to info it slows down the chip and is for development anyway

jaylikesbunda replied:
we could build them in and have max log level to info in the future so that we can change default on the fly at runtime i would assume

tototo31 replied:
Fixed - reverted log level to warning in https://github.com/jaylikesbunda/Ghost_ESP/pull/185/commits/9825a47b48c26daa212edf2c2b2d39dc70459143

Commit:
revertt logging level to warning

tototo31 committed:
Merge branch 'Development-deki' into ethernet-support

tototo31 committed:
Merge branch 'Development-deki' into ethernet-support

Commit:
Enable LWIP statistics and reorganize USB-OTG configuration in sdkconfig.somethingsomething2

tototo31 commented:
fixed eth stats

tototo31 commented:
cmake was not actually an issue

tototo31 commented:
@jaylikesbunda i think this is ready to be merged when you have the time to re review

tototo31 committed:
Merge branch 'Development-deki' into ethernet-support

Commit:
Enhance ethernet_manager: Improve GPIO ISR handling and link status checking. Added handling for already installed GPIO ISR service and refined link status retrieval logic with PHY register checks.

Commit:
Add ensure_eth_interface_up function to validate Ethernet initialization before command execution

This new helper function checks if the Ethernet interface is up and initializes it if necessary. It includes waiting for link establishment and DHCP assignment, enhancing the reliability of Ethernet-related commands.

Commit:
Add ethntp command for NTP server synchronization

Introduced the `ethntp` command to query an NTP server and synchronize the system time. The command allows specifying a custom NTP server, with a default of `pool.ntp.org`. Updated documentation and help output to include usage examples and requirements for an active Ethernet connection.

Commit:
Add ethhttp command for HTTP/HTTPS requests

Introduced the `ethhttp` command to send HTTP/HTTPS GET requests and display the server response. The command supports optional parameters for limiting the number of displayed lines or showing the full response. Updated documentation to include usage examples and requirements for an active Ethernet connection.

Commit:
Add HTTP Request and Sync NTP Time options to options screen

Enhanced the options screen by adding new functionalities for "HTTP Request" and "Sync NTP Time". The "HTTP Request" option allows users to input a URL for sending HTTP requests, while the "Sync NTP Time" option facilitates synchronization with an NTP server. Updated the corresponding callback functions to handle user input and command execution.

Commit:
Add settime and time commands for system time management

Implemented the `settime` command to allow users to set the system time using a Unix timestamp, along with the `time` command to display the current local and UTC time. Enhanced the NTP synchronization process to send the synced time to a connected peer if dual communication is established. Updated command registration to include the new functionalities.

jaylikesbunda committed:
Merge branch 'Development-deki' into ethernet-support

jaylikesbunda committed:
send stop to peer when leaving ghostlink terminal

jaylikesbunda committed:
Merge branch 'Development-deki' into ethernet-support

tototo31 committed:
Merge branch 'Development-deki' into ethernet-support

jaylikesbunda committed:
Merge branch 'Development-deki' into ethernet-support

jaylikesbunda committed:
Merge branch 'Development-deki' into ethernet-support

jaylikesbunda committed:
minor improvements

jaylikesbunda committed:
wording change

jaylikesbunda committed:
add ethernet fingerprinting

jaylikesbunda committed:
add more eth docs
